uint/src/uint.rs

```diff
@@ -1550,31 +1576,34 @@ macro_rules! construct_uint {
 			}
 		}
 
-		$crate::impl_std_for_uint!($name, $n_words);
-		// `$n_words * 8` because macro expects bytes and
-		// uints use 64 bit (8 byte) words
-		$crate::impl_quickcheck_arbitrary_for_uint!($name, ($n_words * 8));
-		$crate::impl_arbitrary_for_uint!($name, ($n_words * 8));
-	}
-}
-
-#[cfg(feature = "std")]
-#[macro_export]
-#[doc(hidden)]
-macro_rules! impl_std_for_uint {
-	($name: ident, $n_words: tt) => {
 		impl $crate::core_::str::FromStr for $name {
-			type Err = $crate::rustc_hex::FromHexError;
+			type Err = $crate::FromHexError;
 
 			fn from_str(value: &str) -> $crate::core_::result::Result<$name, Self::Err> {
-				use $crate::rustc_hex::FromHex;
-				let bytes: Vec<u8> = match value.len() % 2 == 0 {
-					true => value.from_hex()?,
-					false => ("0".to_owned() + value).from_hex()?,
-				};
+				const BYTES_LEN: usize = $n_words * 8;
+				const MAX_ENCODED_LEN: usize = BYTES_LEN * 2;
+
+				let mut bytes = [0_u8; BYTES_LEN];
+
+				let encoded = value.as_bytes();
 
-				if $n_words * 8 < bytes.len() {
-					return Err(Self::Err::InvalidHexLength);
+				if encoded.len() > MAX_ENCODED_LEN {
```

Comment:
I find this code hard to read, it could maybe be possible to refactor to avoid repeating `let out = &mut bytes[BYTES_LEN - encoded.len() / 2..];` and `$crate::hex::decode_to_slice(encoded, out).map_err(Self::Err::from)?;` twice.

Reply:
I'm not sure it's possible because owning buffers are of different types (`value: &str` vs `s: [u8; MAX_ENCODED_LEN]`), different lengths and borrowing is involved here.